sdk/python/kfp/client/client.py

```diff
@@ -1666,3 +1627,71 @@ def validate_pipeline_resource_name(name: str) -> None:
         raise ValueError(
             f'Invalid pipeline name: "{name}". Pipeline name must conform to the regex: "{REGEX}".'
         )
+
+
+def _extract_pipeline_yaml(package_file: str) -> _PipelineDoc:
+
+    def _choose_pipeline_file(file_list: List[str]) -> str:
+        pipeline_files = [file for file in file_list if file.endswith('.yaml')]
+        if not pipeline_files:
+            raise ValueError(
+                'Invalid package. Missing pipeline yaml file in the package.')
+
+        if 'pipeline.yaml' in pipeline_files:
+            return 'pipeline.yaml'
+        elif len(pipeline_files) == 1:
+            return pipeline_files[0]
+        else:
+            raise ValueError(
+                'Invalid package. There is no pipeline.json file or there '
+                'are multiple yaml files.')
+
+    def _safe_load_yaml(stream: TextIO) -> _PipelineDoc:
+        docs = yaml.safe_load_all(stream)
+        pipeline_spec_dict = None
+        platform_spec_dict = {}
+        for doc in docs:
+            if pipeline_spec_dict is None:
+                pipeline_spec_dict = doc
+            else:
+                platform_spec_dict.update(doc)
```

Comment:
nit: Would this logic be simpler and equivalent if we check if `len(list(docs))` is 1 or 2, then construct `_PipelineDoc`? Something like [this](https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/components/structures.py#L1052). IIUC, this is done implicitly in the current implementation via the expectation that `pipeline_spec_dict` will be `None` on iteration 1 and `not None` on iteration 2.

Reply:
I was thinking it could be more than 2, so I implemented a dict merge logic for platform configs spread across multi docs, e.g.:
```yaml
<IR>
---
platforms:
    kubernetes:
        ...
---
platforms:
   vertex:
       ...
---
platforms:
    aws:
       ....
```
This should be a legitimate structure if we consider the "overlay"-like concept. So I should probably change [this](https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/components/structures.py#L1052) as well. WDYT?

Reply:
Okay, I guessing per current compiler implementation, it would only produce two docs even there're multiple platforms:
```yaml
<IR>
---
platforms:
   kubernetes:
        ...
   vertex:
       ...
   aws:
       ....
```
I'm thinking maybe in future we should consider the more than 2 docs structure. The benefits in my opinion is if I took a platform template, only care about one platform config, it's slightly easier to strip off other platform configs.

Reply:
SGTM. I don't see a strong need for this currently, but I'm definitely open to permitting >1 platform config document in the future.

I think there is probably a slight maintenance cost associated with having more than one way to represent the same information, but this cost is probably minimal.

Reply:
SG. I'll keep both loading code as-is for now. Both can handle 2 docs correctly at this moment, and once we have a clear needs on whether we need multi-doc support, we may consolidate the two methods.

Reply:
SGTM 👍🏼